types/wlr_xdg_foreign_v1.c

```diff
@@ -0,0 +1,458 @@
+}
+
+static void disconnect_imported(struct wlr_xdg_imported_v1 *imported) {
+	if (imported->exported != NULL) {
+		imported->exported = NULL;
+		zxdg_imported_v1_send_destroyed(imported->resource);
+
+		wl_list_remove(&imported->exported_destroyed.link);
+		wl_list_init(&imported->exported_destroyed.link);
+	}
+}
+
+static void xdg_exported_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_xdg_exported_v1 *exported =
+		xdg_exported_from_resource(resource);
+
+	wlr_signal_emit_safe(&exported->base.events.destroy, NULL);
+	wlr_xdg_foreign_exported_finish(&exported->base);
+
+	wl_list_remove(&exported->xdg_surface_unmap.link);
+	wl_list_remove(&exported->link);
+	free(exported);
+}
+
+static void handle_xdg_surface_unmap(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_exported_v1 *exported =
+		wl_container_of(listener, exported, xdg_surface_unmap);
+	wl_resource_destroy(exported->resource);
+}
+
+static void xdg_exporter_handle_export(struct wl_client *wl_client,
+		struct wl_resource *client_resource,
+		uint32_t id,
+		struct wl_resource *surface_resource) {
+	struct wlr_xdg_exporter_v1 *exporter =
+		xdg_exporter_from_resource(client_resource);
+	struct wlr_surface *surface = wlr_surface_from_resource(surface_resource);
+
+	if (!verify_is_toplevel(client_resource, surface)) {
+		return;
+	}
+
+	struct wlr_xdg_exported_v1 *exported =
+		calloc(1, sizeof(struct wlr_xdg_exported_v1));
+	if (exported == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	if (!wlr_xdg_foreign_exported_init(&exported->base, exporter->foreign->registry)) {
+		wl_client_post_no_memory(wl_client);
+		free(exported);
+		return;
+	}
+
+	exported->base.surface = surface;
+	exported->resource = wl_resource_create(wl_client,
+		&zxdg_exported_v1_interface,
+		wl_resource_get_version(exporter->resource),
+		id);
+
+	wl_resource_set_implementation(exported->resource, &xdg_exported_impl,
+			exported, xdg_exported_handle_resource_destroy);
+
+	wl_list_insert(&exporter->exports, &exported->link);
+	wl_list_init(&exported->imports);
+	wl_signal_init(&exported->base.events.destroy);
+
+	zxdg_exported_v1_send_handle(exported->resource, exported->base.handle);
+
+	exported->xdg_surface_unmap.notify = handle_xdg_surface_unmap;
+	if (wlr_surface_is_xdg_surface(surface)) {
+		struct wlr_xdg_surface *xdg_surface =
+			wlr_xdg_surface_from_wlr_surface(surface);
+		wl_signal_add(&xdg_surface->events.unmap, &exported->xdg_surface_unmap);
+	}
+}
+
+static const struct zxdg_exporter_v1_interface xdg_exporter_impl = {
+	.destroy = xdg_exporter_handle_destroy,
+	.export = xdg_exporter_handle_export
+};
+
+static void xdg_exporter_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_xdg_exporter_v1 *exporter = xdg_exporter_from_resource(resource);
+	struct wlr_xdg_exported_v1 *exported, *exported_tmp;
+	wl_list_for_each_safe(exported, exported_tmp, &exporter->exports, link) {
+		wl_resource_destroy(exported->resource);
```

Comment:
Nothing in the protocol says that destroying the exporter also destroys the child objects.